### DISCUSSION TITLE
Due TruLens work with Local LLM using Langchain without OpenAI Key?

### DISCUSSION THREAD

**Discussion by navinchandrac** [truera/trulens#673] [Q&A] [answered]
Does TruLens work with Local LLM using Langchain without OpenAI Key? (Example for Local LLM: https://colab.research.google.com/drive/1XVjur9cdYLhxe6mrQwQUlWIm4VJmiYlu?usp=sharing#scrollTo=wJ7k5mYXj0vG)

Getting below error while using TruLens with Local LLM using Langchain for above example

**_ImportError: cannot import name 'LLM' from 'llama_index.llms.base' (/usr/local/lib/python3.10/dist-packages/llama_index/llms/base.py)_**

Also which LLM used by TruLens (as internal) for to get expected feedback parameters result comparison with LLM App actual feedback result?

**Comment by jimwhite** [comment, marked as answer]
You don't show your code so the error doesn't make much sense because it references LlamaIndex but the notebook you link appears to use LangChain.  

The LiteLLM class provides compatibility with various LLM implementations: https://www.trulens.org/trulens_eval/api/litellm_provider/

Here's a snippet of how I use it with Gemini (and LlamaIndex):

```python
from trulens_eval import Feedback, LiteLLM, TruLlama
from trulens_eval.feedback import Groundedness

# Initialize provider class
GEMINI_PROVIDER = LiteLLM(model_engine="gemini-pro")

GROUNDED = Groundedness(groundedness_provider=GEMINI_PROVIDER)

# Define a groundedness feedback function
f_groundedness = (
    Feedback(GROUNDED.groundedness_measure_with_cot_reasons)
    .on(TruLlama.select_source_nodes().node.text.collect())
    .on_output()
    .aggregate(GROUNDED.grounded_statements_aggregator)
)
```

**Reply by YasaminAbbaszadegan** [reply]
I installed "pip install "litellm<=1.24.0,>=1.11.1" but it gives me this error--> 
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[10], [line 2](vscode-notebook-cell:?execution_count=10&line=2)
      [1](vscode-notebook-cell:?execution_count=10&line=1) # Initialize provider class
----> [2](vscode-notebook-cell:?execution_count=10&line=2) GEMINI_PROVIDER = LiteLLM(model_engine=llm)
      [4](vscode-notebook-cell:?execution_count=10&line=4) GROUNDED = Groundedness(groundedness_provider="gemini-pro")
      [6](vscode-notebook-cell:?execution_count=10&line=6) # Define a groundedness feedback function

File [~/.virtualenvs/venv/lib/python3.10/site-packages/trulens_eval/utils/imports.py:364](https://file+.vscode-resource.vscode-cdn.net/Users/yasaminabbaszadegan/Desktop/project/LLM_RAG/RAG_Toronto_Zoning_Notices/~/.virtualenvs/venv/lib/python3.10/site-packages/trulens_eval/utils/imports.py:364), in Dummy.__call__(self, *args, **kwargs)
    [363](https://file+.vscode-resource.vscode-cdn.net/Users/yasaminabbaszadegan/Desktop/project/LLM_RAG/RAG_Toronto_Zoning_Notices/~/.virtualenvs/venv/lib/python3.10/site-packages/trulens_eval/utils/imports.py:363) def __call__(self, *args, **kwargs):
--> [364](https://file+.vscode-resource.vscode-cdn.net/Users/yasaminabbaszadegan/Desktop/project/LLM_RAG/RAG_Toronto_Zoning_Notices/~/.virtualenvs/venv/lib/python3.10/site-packages/trulens_eval/utils/imports.py:364)     raise self.exception_class(self.message)

ModuleNotFoundError: 
litellm package is required for using LiteLLM models.
You should be able to install it with pip:

    ```bash
    pip install "litellm<=1.24.0,>=1.11.1"
    ```

**Comment by kinthaiofficial** [comment]
TruLens with local LLMs is definitely doable — the main thing to configure is the feedback function provider and the completion endpoint.

For Langchain + local model:

```python
from trulens_eval import TruChain, Feedback, Tru
from trulens_eval.feedback.provider import OpenAI as OpenAIProvider

# Point to Ollama's OpenAI-compatible endpoint
from openai import OpenAI
client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")

provider = OpenAIProvider(client=client, model_engine="llama3")

# Build feedback functions using local model as judge
f_relevance = Feedback(provider.relevance).on_input_output()
f_groundedness = Feedback(provider.groundedness_measure_with_cot_reasons).on(...)
```

A few caveats:

**Local model as judge quality** — feedback functions like groundedness, relevance, and answer correctness require the judging model to follow nuanced instructions. 70B models (Llama 3 70B, Qwen 2.5 72B) work well as judges; 7B-13B models can be inconsistent and will require recalibration.

**Consistency sampling** — local models tend to be less consistent than frontier models on the same judgment task. Sample each feedback function 3 times and take the majority/average for more reliable scores.

**Template adaptation** — TruLens's feedback prompts are optimized for GPT-4 instruction following. Some prompts may need minor adaptation for other model families, especially around JSON output format.

For agent pipeline evaluation specifically, there's a write-up on the evaluation dimensions that matter for multi-agent systems: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

Are you evaluating RAG quality, agent decision quality, or something else?
